File: AssetManager/UserInterface/Forms/AdminTools/CopyFilesForm.vb
Public Class CopyFilesForm
    Private TargetDevice As DeviceObject
    Private SourceDirectory As String
    Private TargetDirectory As String
    Private PushFilesControl As GKProgressControl
    Private Cancel As Boolean = False
    Sub New(parentForm As ExtendedForm, targetDevice As DeviceObject, sourceDirectory As String, targetDirectory As String)
        InitializeComponent()
        ImageCaching.CacheControlImages(Me)
        Me.ParentForm = parentForm
        Me.TargetDevice = targetDevice
        Me.SourceDirectory = sourceDirectory
        Me.TargetDirectory = targetDirectory
        PushFilesControl = New GKProgressControl(Me, targetDevice, True, sourceDirectory, targetDirectory)
        Me.Controls.Add(PushFilesControl)

        Me.Show()

    End Sub

    Public Async Function StartCopy() As Task(Of Boolean)
        Try
            PushFilesControl.StartUpdate()

            Dim Done = Await Task.Run(Function()
                                          Do Until PushFilesControl.ProgStatus <> GKProgressControl.ProgressStatus.Running And PushFilesControl.ProgStatus <> GKProgressControl.ProgressStatus.Starting
                                              If Cancel Then Return False
                                              Task.Delay(1000).Wait()
                                          Loop
                                          If PushFilesControl.ProgStatus <> GKProgressControl.ProgressStatus.CompleteWithErrors And PushFilesControl.ProgStatus <> GKProgressControl.ProgressStatus.Complete Then
                                              Return False
                                          Else
                                              Return True
                                          End If
                                      End Function)
            Return Done
        Catch ex As Exception
            Debug.Print(ex.Message)
            Return False
        End Try

    End Function

    Private Sub CopyFilesForm_Disposed(sender As Object, e As EventArgs) Handles Me.Disposed
        If PushFilesControl IsNot Nothing Then PushFilesControl.Dispose()
        Cancel = True
    End Sub
End Class

File: AssetManager/UserInterface/Forms/AdminTools/CopyFilesForm.cs
using System;
using System.Threading.Tasks;
using AssetManager.UserInterface.Forms.GK_Updater;
using AssetManager.UserInterface.CustomControls;



namespace AssetManager.UserInterface.Forms.AdminTools
{
    public partial class CopyFilesForm : ExtendedForm
    {
        private DeviceObject TargetDevice;
        private string SourceDirectory;
        private string TargetDirectory;
        private GKProgressControl PushFilesControl;
        private bool Cancel = false;
        public CopyFilesForm(ExtendedForm parentForm, DeviceObject targetDevice, string sourceDirectory, string targetDirectory)
        {
            InitializeComponent();
            ImageCaching.CacheControlImages(this);
            this.ParentForm = parentForm;
            this.Owner = parentForm;
            this.TargetDevice = targetDevice;
            this.SourceDirectory = sourceDirectory;
            this.TargetDirectory = targetDirectory;
            PushFilesControl = new GKProgressControl(this, targetDevice, true, sourceDirectory, targetDirectory);
            this.Controls.Add(PushFilesControl);
            PushFilesControl.CriticalStopError += new System.EventHandler(CopyCritcalError);
            this.Show();

        }

        public async Task<bool> StartCopy()
        {
            try
            {
                PushFilesControl.StartUpdate();

                var Done = await Task.Run(() =>
                {
                    while (!(PushFilesControl.ProgStatus != GKProgressControl.ProgressStatus.Running && PushFilesControl.ProgStatus != GKProgressControl.ProgressStatus.Starting))
                    {
                        if (Cancel)
                        {
                            PushFilesControl.CancelUpdate();
                            return false;
                        }
                        Task.Delay(1000).Wait();
                    }
                    if (PushFilesControl.ProgStatus != GKProgressControl.ProgressStatus.CompleteWithErrors && PushFilesControl.ProgStatus != GKProgressControl.ProgressStatus.Complete)
                    {
                        return false;
                    }
                    else
                    {
                        return true;
                    }
                });
                return Done;
            }
            catch (Exception ex)
            {
                throw (ex);
                return false;
            }

        }

        private void CopyCritcalError(object sender, EventArgs e)
        {
            Cancel = true;
            SecurityTools.ClearAdminCreds();
        }

        private void CopyFilesForm_FormClosed(object sender, System.Windows.Forms.FormClosedEventArgs e)
        {
            if (PushFilesControl != null)
            {
                PushFilesControl.Dispose();
            }
            Cancel = true;
        }
    }
}
